From: Doug Ledford <dledford@redhat.com>
Date: Tue, 1 Dec 2009 19:33:15 -0500
Subject: [scsi] sg: rate limit warning
Message-id: <1259696003-21028-1-git-send-email-dledford@redhat.com>
Patchwork-id: 21584
O-Subject: [Patch RHEL5.5] [sg] rate limit warning
Bugzilla: 536937
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Jarod Wilson <jarod@redhat.com>
RH-Acked-by: Jiri Olsa <jolsa@redhat.com>

Taken from upstream commit eaa3e22e8d32bf7a6176f04efad90f4a5aa67f58
Resolves bz536937

Signed-off-by: Doug Ledford <dledford@redhat.com>

diff --git a/drivers/scsi/sg.c b/drivers/scsi/sg.c
index 79ecf4e..1b94762 100644
--- a/drivers/scsi/sg.c
+++ b/drivers/scsi/sg.c
@@ -606,8 +606,9 @@ sg_write(struct file *filp, const char __user *buf, size_t count, loff_t * ppos)
 	 * but is is possible that the app intended SG_DXFER_TO_DEV, because there
 	 * is a non-zero input_size, so emit a warning.
 	 */
-	if (hp->dxfer_direction == SG_DXFER_TO_FROM_DEV)
-		if (printk_ratelimit())
+	if (hp->dxfer_direction == SG_DXFER_TO_FROM_DEV) {
+		static char cmd[TASK_COMM_LEN];
+		if (strcmp(current->comm, cmd) && printk_ratelimit()) {
 			printk(KERN_WARNING
 			       "sg_write: data in/out %d/%d bytes for SCSI command 0x%x--"
 			       "guessing data in;\n" KERN_WARNING "   "
@@ -615,6 +616,9 @@ sg_write(struct file *filp, const char __user *buf, size_t count, loff_t * ppos)
 			       old_hdr.reply_len - (int)SZ_SG_HEADER,
 			       input_size, (unsigned int) cmnd[0],
 			       current->comm);
+			strcpy(cmd, current->comm);
+		}
+	}
 	k = sg_common_write(sfp, srp, cmnd, sfp->timeout, blocking);
 	return (k < 0) ? k : count;
 }
